### PR TITLE
[SR SQL Planner Summer Camp] New Parser and Analyzer support ShowLoadStmt/ShowLoadWarnin…

### LINKED ISSUE #7955: New Parser and Analyzer support Load related-statements (Challenging)
# Required work and output

1.  Use antlr4 to complete parse statements. The generated statement must follow the original data structure, and should be compatible with the old statement.
    
2.  In the original syntax file  **sql_parser.cup**, the new syntax needs to be compatible with the syntax of the old version, using the syntax file sql_parser.cup of the old version as the reference standard.
    
3.  Remove the logic of analytic functions in the statement and move the relevant logic to the new Analyzer framework. The new Analyzer is in the SQL/Analyzer directory.
    
4.  If the code related to Privilege Check exists in the original analyzer, move the code related to privilege control into PrivilegeChecker.
    
5.  When submitting the PR, please make sure there is unit test coverage for related changes.
    

# Tasks

 - [ ] ShowLoadStmt
 - [ ] CancelLoadStmt
 - [ ] ShowLoadWarningsStmt
 - [ ] LoadStmt
 - [ ] ModifyBrokerClause
 - [ ] ShowBrokerStmt

# How to participate?

1.  Pick an issue you are interested in, click the issue link and leave a comment under the issue to sign up.  **One person can only choose one issue and the first person who picks the issue will be assigned.**
    
2.  **After the issue is assigned, you should complete the task in 14 days.** If you are unable to complete it within the specified time, the task will be withdrawn and assigned to others.
    
3.  Participants are required to produce two pieces of content (forms including but not limited to mind maps, study notes, blogs, and tutorials) in the course of the activity, each of which should be at least 200 words.
    
4.  Got questions? Join our  [Slack channel](https://join.slack.com/t/starrocks/shared_invite/zt-192zeqlc7-9MAvqAXNcveW_l5Pqr8r2g)  and find channel:  **#sql-planner**.
    

# Before you start

1.  Read  [Contributing.md](https://github.com/StarRocks/starrocks/blob/main/CONTRIBUTING.md)  carefully to understand how to submit a pull request and how to configure BE and FE development environment.
    
2.  Sign the  [Contributor License Agreement](https://cla-assistant.io/StarRocks/starrocks)  before you make a pull request.
    
3.  To learn more about StarRocks, please refer to  [StarRocks documentation](https://github.com/StarRocks/docs/blob/main/TOC.md).
    

# Pull request requirements

1.  Create a pull request (PR) after you complete the task.
* PR title format: [SR SQL Planner Summer Camp] + PR title
2. After submitting PR, you can comment under the pinned Issue [SR SQL Planner Summer Camp].
* Comment format: "I have finished Issue #" + PR link

### PR DESCRIPTION
**Pull request by predator4ann**
…gsStmt/CancelLoadStmt

## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：
Fixes #7955

## Problem Summary(Required) ：
New Parser and Analyzer support ShowLoadStmt/ShowLoadWarningsStmt/CancelLoadStmt

### CODE REVIEW

**Comment by predator4ann** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 248 / 319 (77.74%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|2|3|66.67%|[304]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowLoadWarningsStmtAnalyzer.java|63|92|68.48%|[23, 58, 62, 69, 78, 89, 100, 104, 105, 106, 110, 111, 112, 120, 121, 127, 128, 131, 132, 137, 138, 152, 153, 158, 159, 167, 168, 169, 178]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CancelLoadStmtAnalyzer.java|33|45|73.33%|[16, 42, 62, 63, 66, 67, 72, 73, 82, 83, 88, 94]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowLoadStmtAnalyzer.java|79|102|77.45%|[26, 57, 81, 88, 99, 105, 123, 124, 129, 130, 133, 134, 139, 140, 148, 149, 153, 154, 163, 164, 169, 170, 188]|
|:large_blue_circle:|com/starrocks/analysis/CancelLoadStmt.java|8|10|80.00%|[58, 68]|
|:large_blue_circle:|com/starrocks/analysis/ShowLoadStmt.java|13|16|81.25%|[84, 143, 152]|
|:large_blue_circle:|com/starrocks/analysis/ShowLoadWarningsStmt.java|12|13|92.31%|[131]|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|34|34|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowStmtAnalyzer.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 278 / 319 (87.15%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|2|3|66.67%|[304]|
|:large_blue_circle:|com/starrocks/analysis/CancelLoadStmt.java|8|10|80.00%|[58, 68]|
|:large_blue_circle:|com/starrocks/analysis/ShowLoadStmt.java|13|16|81.25%|[84, 143, 152]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowLoadWarningsStmtAnalyzer.java|76|92|82.61%|[23, 58, 62, 69, 78, 89, 100, 106, 110, 111, 112, 120, 121, 137, 138, 178]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowLoadStmtAnalyzer.java|89|102|87.25%|[26, 57, 81, 88, 99, 105, 129, 130, 139, 140, 148, 149, 188]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CancelLoadStmtAnalyzer.java|40|45|88.89%|[16, 42, 72, 73, 94]|
|:large_blue_circle:|com/starrocks/analysis/ShowLoadWarningsStmt.java|12|13|92.31%|[131]|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|34|34|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowStmtAnalyzer.java|2|2|100.00%|[]|

**Comment by predator4ann** [comment]
run starrocks_admit_test

**Comment by predator4ann** [comment]
cc @HangyuanLiu @imay PTAL

**Review by HangyuanLiu — DISMISSED** [review]

**Comment by HangyuanLiu** [comment]
Please rebase, and I will merge PR as soon as possible

**Comment by liuyehcf** [comment]
This is a huge pr. Better to split this pr into 3 sub prs, ShowLoadStmt, ShowLoadStmt, CancelLoadStmt respectively.

**Comment by HangyuanLiu** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 277 / 317 (87.38%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|2|3|66.67%|[313]|
|:large_blue_circle:|com/starrocks/analysis/CancelLoadStmt.java|8|10|80.00%|[58, 68]|
|:large_blue_circle:|com/starrocks/analysis/ShowLoadStmt.java|13|16|81.25%|[84, 143, 152]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowLoadWarningsStmtAnalyzer.java|75|90|83.33%|[23, 58, 62, 69, 78, 95, 103, 107, 108, 109, 117, 118, 134, 135, 175]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowLoadStmtAnalyzer.java|89|102|87.25%|[26, 57, 81, 88, 99, 105, 129, 130, 139, 140, 148, 149, 188]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CancelLoadStmtAnalyzer.java|40|45|88.89%|[16, 42, 72, 73, 94]|
|:large_blue_circle:|com/starrocks/analysis/ShowLoadWarningsStmt.java|12|13|92.31%|[131]|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|34|34|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowStmtAnalyzer.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 277 / 317 (87.38%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|2|3|66.67%|[313]|
|:large_blue_circle:|com/starrocks/analysis/CancelLoadStmt.java|8|10|80.00%|[58, 68]|
|:large_blue_circle:|com/starrocks/analysis/ShowLoadStmt.java|13|16|81.25%|[84, 143, 152]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowLoadWarningsStmtAnalyzer.java|75|90|83.33%|[23, 58, 62, 69, 78, 95, 103, 107, 108, 109, 117, 118, 134, 135, 175]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowLoadStmtAnalyzer.java|89|102|87.25%|[26, 57, 81, 88, 99, 105, 129, 130, 139, 140, 148, 149, 188]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CancelLoadStmtAnalyzer.java|40|45|88.89%|[16, 42, 72, 73, 94]|
|:large_blue_circle:|com/starrocks/analysis/ShowLoadWarningsStmt.java|12|13|92.31%|[131]|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|34|34|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowStmtAnalyzer.java|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ShowLoadStmtAnalyzer.java [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+import com.starrocks.qe.ConnectContext;
+import com.starrocks.sql.ast.AstVisitor;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class ShowLoadStmtAnalyzer {
+
+    public static Void analyze(ShowLoadStmt statement, ConnectContext context) {
+        return new ShowLoadStmtAnalyzerVisitor().visitShowLoadStmt(statement, context);
```
Comment (liuyehcf) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ShowLoadStmtAnalyzer.java:29]: In visitor style, should call `visit()` here

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ShowLoadWarningsStmtAnalyzer.java [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+
+public class ShowLoadWarningsStmtAnalyzer {
+
+    public static Void analyze(ShowLoadWarningsStmt statement, ConnectContext context) {
+        return new ShowLoadWarningsStmtAnalyzerVisitor().visitShowLoadWarningsStmt(statement, context);
```
Comment (liuyehcf) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ShowLoadWarningsStmtAnalyzer.java:26]: should call `visit()` here

File: fe/fe-core/src/main/java/com/starrocks/analysis/CancelLoadStmt.java [hunk 3]
```diff
@@ -34,82 +32,48 @@ public String getDbName() {
...
+        return true;
     }
 
     @Override
     public String toSql() {
         StringBuilder stringBuilder = new StringBuilder();
         stringBuilder.append("CANCEL LOAD ");
         if (!Strings.isNullOrEmpty(dbName)) {
-            stringBuilder.append("FROM " + dbName);
+            stringBuilder.append("FROM `").append(dbName).append("`");
```
[the remark this answers is not in the document]
Comment (predator4ann) [reply fe/fe-core/src/main/java/com/starrocks/analysis/CancelLoadStmt.java:76]: dbName like default_cluster:test without backquote wrapped will cause syntax error, so changed it liking ShowLoadStmt#toSql()

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ShowLoadStmtAnalyzer.java (hunk 1 above)
Comment (predator4ann) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ShowLoadStmtAnalyzer.java]: done

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ShowLoadWarningsStmtAnalyzer.java (hunk 2 above)
Comment (predator4ann) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ShowLoadWarningsStmtAnalyzer.java]: done

File: fe/fe-core/src/main/java/com/starrocks/analysis/CancelLoadStmt.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (predator4ann) [reply fe/fe-core/src/main/java/com/starrocks/analysis/CancelLoadStmt.java:76]: and it should be compatible with the old statement
https://github.com/StarRocks/starrocks/blob/dfbf1e9bf258f3a530e2f895dec12f70aec424b2/fe/fe-core/src/main/cup/sql_parser.cup#L994

**Comment by predator4ann** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 284 / 328 (86.59%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|2|3|66.67%|[314]|
|:large_blue_circle:|com/starrocks/analysis/CancelLoadStmt.java|8|10|80.00%|[58, 68]|
|:large_blue_circle:|com/starrocks/analysis/ShowLoadStmt.java|13|16|81.25%|[84, 143, 152]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowLoadWarningsStmtAnalyzer.java|76|93|81.72%|[23, 42, 43, 62, 66, 73, 82, 99, 107, 111, 112, 113, 121, 122, 138, 139, 179]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowLoadStmtAnalyzer.java|90|105|85.71%|[26, 45, 46, 61, 85, 92, 103, 109, 133, 134, 143, 144, 152, 153, 192]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CancelLoadStmtAnalyzer.java|43|48|89.58%|[16, 46, 76, 77, 98]|
|:large_blue_circle:|com/starrocks/analysis/ShowLoadWarningsStmt.java|12|13|92.31%|[131]|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|34|34|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowStmtAnalyzer.java|4|4|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 281 / 325 (86.46%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|2|3|66.67%|[314]|
|:large_blue_circle:|com/starrocks/analysis/CancelLoadStmt.java|8|10|80.00%|[58, 68]|
|:large_blue_circle:|com/starrocks/analysis/ShowLoadStmt.java|13|16|81.25%|[84, 143, 152]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowLoadWarningsStmtAnalyzer.java|75|92|81.52%|[22, 41, 42, 61, 65, 72, 81, 96, 104, 108, 109, 110, 118, 119, 135, 136, 176]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowLoadStmtAnalyzer.java|89|104|85.58%|[25, 44, 45, 60, 82, 89, 100, 106, 130, 131, 140, 141, 149, 150, 189]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CancelLoadStmtAnalyzer.java|42|47|89.36%|[15, 45, 73, 74, 95]|
|:large_blue_circle:|com/starrocks/analysis/ShowLoadWarningsStmt.java|12|13|92.31%|[131]|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|34|34|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowStmtAnalyzer.java|4|4|100.00%|[]|

**Comment by predator4ann** [comment]
cc @liuyehcf @HangyuanLiu PTAL

**Comment by predator4ann** [comment]
run starrocks_fe_unittest

**Review by HangyuanLiu — APPROVED** [review]

**Comment by liuyehcf** [comment]
LGTM

**Review by liuyehcf — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Pull request merged**